src/virtio-blk.c

```diff
@@ -0,0 +1,161 @@
+
+        memcpy(&req, vm_guest_to_host(v, (void *) desc->addr), desc->len);
+        if (req.type == VIRTIO_BLK_T_IN || req.type == VIRTIO_BLK_T_OUT) {
+            if (!virtq_check_next(desc))
+                return;
+            desc = virtq_get_avail(vq);
+            req.data_size = desc->len;
+            req.data = vm_guest_to_host(v, (void *) desc->addr);
+
+            ssize_t r;
+            if (req.type == VIRTIO_BLK_T_IN)
+                r = virtio_blk_read(dev, req.data, req.sector << 9,
+                                    req.data_size);
+            else
+                r = virtio_blk_write(dev, req.data, req.sector << 9,
+                                     req.data_size);
+
+            status = r < 0 ? VIRTIO_BLK_S_IOERR : VIRTIO_BLK_S_OK;
+        } else {
+            status = VIRTIO_BLK_S_UNSUPP;
+        }
+        if (!virtq_check_next(desc))
+            return;
+        desc = virtq_get_avail(vq);
+        req.status = vm_guest_to_host(v, (void *) desc->addr);
+        *req.status = status;
+        used_desc->flags ^= (1ULL << VRING_PACKED_DESC_F_USED);
+        used_desc->len = r;
+        dev->virtio_pci_dev.config.isr_cap.isr_status |= VIRTIO_PCI_ISR_QUEUE;
+    }
+}
+
+static struct virtq_ops ops = {
+    .enable_vq = virtio_blk_enable_vq,
+    .complete_request = virtio_blk_complete_request,
+    .notify_used = virtio_blk_notify_used,
+};
+
+void virtio_blk_init_pci(struct virtio_blk_dev *virtio_blk_dev,
+                         struct diskimg *diskimg,
+                         struct pci *pci,
+                         struct bus *io_bus,
+                         struct bus *mmio_bus)
+{
+    virtio_blk_init(virtio_blk_dev);
+    virtio_blk_dev->diskimg = diskimg;
+    virtio_blk_dev->config.capacity = diskimg->size >> 9;
+    struct virtio_pci_dev *dev = &virtio_blk_dev->virtio_pci_dev;
+    virtio_pci_init(dev, pci, io_bus, mmio_bus);
+    virtio_pci_set_dev_cfg(dev, &virtio_blk_dev->config,
+                           sizeof(virtio_blk_dev->config));
+    virtio_pci_set_pci_hdr(dev, VIRTIO_PCI_DEVICE_ID_BLK, VIRTIO_BLK_PCI_CLASS,
+                           virtio_blk_dev->irq_num);
+    virtio_pci_set_virtq(dev, virtio_blk_dev->vq, VIRTIO_BLK_VIRTQ_NUM);
+    virtio_pci_add_feature(dev, 0);
+    virtio_pci_enable(dev);
+}
+
+void virtio_blk_init(struct virtio_blk_dev *dev)
+{
+    vm_t *v = container_of(dev, vm_t, virtio_blk_dev);
+
+    memset(dev, 0x00, sizeof(struct virtio_blk_dev));
+    dev->ioeventfd = eventfd(0, EFD_CLOEXEC);
+    dev->irqfd = eventfd(0, EFD_CLOEXEC);
+    /* FIXME: irq_num should be different to other devs */
+    dev->irq_num = 15;
+    vm_irqfd_register(v, dev->irqfd, dev->irq_num, 0);
+    dev->diskimg = NULL;
+    for (int i = 0; i < VIRTIO_BLK_VIRTQ_NUM; i++)
+        virtq_init(&dev->vq[i], dev, &ops);
+}
+
+void virtio_blk_exit(struct virtio_blk_dev *dev)
+{
+    if (dev->diskimg)
+        diskimg_exit(dev->diskimg);
+    virtio_pci_exit(&dev->virtio_pci_dev);
```

Comment:
`virtio_blk_exit` will be executed even if we don't input the disk image, but `dev->diskimg` is uninitialized at that time. We should not access it in this case.